File: pkg/util/fzf.go
package util

import (
	"bytes"
	"fmt"
	"io"
	"log/slog"
	"os"
	"os/exec"
	"strings"
)

type FzfLine[V any] struct {
	Key    string
	Pretty []string
	Raw    V
}

func FzfGetFromChan[K comparable, V any](
	choices <-chan K,
	fn func(K) (FzfLine[V], error),
	arg ...string,
) (V, error) {
	arg = append([]string{"--with-nth", "2.."}, arg...)

	fzf := exec.Command("fzf", arg...)
	mapping := make(map[string]V)
	out := make(chan V, 1)
	var zero V

	pipeIn, err := fzf.StdinPipe()
	if err != nil {
		slog.Error("Failed to create stdin pipe of fzf", "error", err)
		return zero, err
	}

	pipeOut, err := fzf.StdoutPipe()
	if err != nil {
		slog.Error("Failed to create stdout pipe of fzf", "error", err)
		return zero, err
	}

	// fzf's stderr to this process
	fzf.Stderr = os.Stderr

	go func() {
		defer pipeIn.Close()
		for choice := range choices {

			line, err := fn(choice)
			if err != nil {
				continue
			}
			mapping[line.Key] = line.Raw
			if _, err := fmt.Fprintf(pipeIn, "%s %s\n", line.Key, strings.Join(line.Pretty, " ")); err != nil {
				slog.Error("Failed to write to fzf stdin", "error", err)
				return
			}
		}

	}()

	go func() {
		output, err := io.ReadAll(pipeOut)
		if err != nil {
			slog.Error("Failed to run fzf", "error", err)
			return
		}

		outBytesFirst, _, _ := bytes.Cut(output, []byte(" "))
		out <- mapping[string(outBytesFirst)]
	}()

	if err := fzf.Run(); err != nil {
		slog.Error("Failed to run fzf", "error", err)
		return zero, err
	}
	return <-out, nil
}

File: src/util/fzf.rs
use std::collections::HashMap;
use std::io::Write;
use std::process::{Command, Stdio};

use anyhow::{Context, Result, bail};

pub struct FzfLine<V> {
    pub key: String,
    pub pretty: Vec<String>,
    pub raw: V,
}

/// Spawn fzf with the given choices and return the selected raw value.
pub fn fzf_select<V>(choices: Vec<FzfLine<V>>, extra_args: &[&str]) -> Result<V> {
    let mut args = vec!["--with-nth", "2.."];
    args.extend_from_slice(extra_args);

    let mut fzf = Command::new("fzf")
        .args(&args)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .context("Failed to spawn fzf")?;

    let mut mapping: HashMap<String, V> = HashMap::new();

    {
        let stdin = fzf.stdin.as_mut().context("Failed to open fzf stdin")?;
        for choice in choices {
            let pretty = choice.pretty.join(" ");
            writeln!(stdin, "{} {}", choice.key, pretty)?;
            mapping.insert(choice.key, choice.raw);
        }
    }

    let output = fzf.wait_with_output().context("Failed to run fzf")?;
    if !output.status.success() {
        bail!("fzf exited with non-zero status");
    }

    let selected = String::from_utf8_lossy(&output.stdout);
    let key = selected.split_whitespace().next().unwrap_or("");

    mapping
        .remove(key)
        .context("Selected key not found in mapping")
}
